Add scenario test for channel extensibility model

khdang opened the pull request:
I've ported the Custom Message Interceptor but it seems that the channel extensibility model doesn't work yet.

Ref https://github.com/dotnet/wcf/issues/623 https://github.com/dotnet/wcf/issues/1170

khdang commented:
@dotnet-bot test outerloop Windows_NT
@dotnet-bot test outerloop selfhost Windows_NT

roncain commented:
@khdang could you rebase and try again? All the Outerloop tests failed in unexpected ways. The selfhost failure has a compile error, meaning you are probably also failing to build under IIS.

roncain commented:
If you try to build tools\SelfHostedWcfService\SelfHostedWcfService.sln you will see many compile errors.  This explains why all OuterLoops fail because neither self-host nor IIS-host can build the server side code.

After rebasing and resolving the merge conflicts reported, please do 'build /p:WithCategories=OuterLoop' to verify the code runs before updating the PR

---
In reply to: [219699069](https://github.com/dotnet/wcf/pull/1171#issuecomment-219699069) [](ancestors = 219699069)

roncain commented on src/System.Private.ServiceModel/tools/SelfHostedWcfService/ChannelExtensibilityService.cs:
```diff
@@ -0,0 +1,66 @@
+using System.ServiceModel;
+using System.ServiceModel.Channels;
+
+namespace WcfService
```
This file is in the wrong place.  All server-side .cs files live under the IISHostedWcfService\App_code folder and are included by the self-host project.  Was there some reason you could not put this in IISHostedWcfService?

khdang replied:
This was not intentional I'll move this to IISHostedWcfService\App_code folder

roncain commented on src/System.Private.ServiceModel/tools/SelfHostedWcfService/SelfHostedWCFService.csproj:
```diff
@@ -57,13 +58,14 @@
     <Compile Include="..\IISHostedWcfService\**\*.cs" />
     <Compile Include="Program.cs" />
     <Compile Include="Properties\AssemblyInfo.cs" />
+    <Compile Include="ChannelExtensibilityService.cs" />
```
As mentioned elsewhere, this file does not belong in this project.

khdang replied:
I will remove this line.

roncain commented on src/System.Private.ServiceModel/tests/Scenarios/Binding/Custom/Interceptingbindingelement.cs:
```diff
@@ -0,0 +1,71 @@
+
+using System.ServiceModel.Channels;
+
+public class InterceptingBindingElement : BindingElement
```
The name of this file is not correctly capitalized (InterceptingbindingElement.cs)

StephenBonikowsky replied:
NIT: To be exact 😄  InterceptingBindingElement to match the same file under App_code.

khdang replied:
I've noticed this and updated in VS but it turns out VS doesn't rename the actual file. Let me do that now.

roncain commented on src/System.Private.ServiceModel/tests/Scenarios/Binding/Custom/AsyncResult.cs:
```diff
@@ -0,0 +1,236 @@
+/// A generic base class for IAsyncResult implementations
+/// that wraps a ManualResetEvent.
+/// </summary>
+abstract class AsyncResult : IAsyncResult
```
Could you briefly describe where these files came from?  Are they copied from a sample?  Product code?  I'm asking about AsyncResult.cs, ChainedAsyncResult.cs, and TimeoutHelper.cs mostly.

khdang replied:
They are from the sample code from https://msdn.microsoft.com/library/ms751495(v=vs.100).aspx

roncain commented on src/System.Private.ServiceModel/tests/Scenarios/Binding/Custom/CustomBindingTests.cs:
```diff
@@ -51,4 +51,69 @@ public static void DefaultSettings_Tcp_Binary_Echo_RoundTrips_String()
+                new HttpTransportBindingElement());
+
+            factory = new ChannelFactory<IChannelExtensibilityContract>(binding, new EndpointAddress(Endpoints.HttpBaseAddress_ChannelExtensibility));
+            serviceProxy = factory.CreateChannel();
```
You opened #1170 and marked this test ActiveIssue with it.  Did you investigate why it failed?  Does it pass if you call factory.Open() first?

khdang replied:
I think we shouldn't need factory.Open() for it to work. In fact, I've tried to call factory.Open() first but it still fails.
I've put an break at where the inner factory is open and that code has never been executed so I think there is something wrong because class InterceptingChannelFactory<TChannel>
.OnOpen() method is never invoked, whereas my understanding is that it should.

roncain commented:
I think the tests really belong in a new folder under Scenarios/Extensibility.  This is all about the extensibility at the channel layer and less about testing CustomBinding.
@StephenBonikowsky -- thoughts on correct placement?

StephenBonikowsky commented:
Yes I agree, this test is meant to replace the old test we had that was based on the Mahjong game to test the MessageInspector scenario.
I think you should place it under Scenarios/Extensibility in a new "MessageInspector" directory.

Also, there are a lot of extra files needed for the test, I would recommend putting the test itself in a MessageInspectorTests.cs and all the other files under another subdirectory called "TestData", I plan on going back and making this same kind of change in other test folders, this way it is easy to differentiate files containing the tests and any other supporting files.

You will need to add a new *.csproj and add it to the solution.

StephenBonikowsky commented on src/System.Private.ServiceModel/tests/Scenarios/Binding/Custom/CustomBindingTests.cs:
```diff
@@ -51,4 +51,69 @@ public static void DefaultSettings_Tcp_Binary_Echo_RoundTrips_String()
     }
+
+    public class MessageModifier : ChannelMessageInterceptor
+    {
```
Test files should only contain test methods, could you move this to a separate file or into one of the other test support files you already have.

I have a task to clean this up in other areas of our test bed.

khdang replied:
Thanks Ron and Stephen! That makes sense. Let me move this new test to Scenarios/Extensibility

khdang commented:
@dotnet-bot test outerloop Windows_NT
@dotnet-bot test outerloop selfhost Windows_NT

roncain commented on src/System.Private.ServiceModel/tests/Scenarios/Extensibility/MessageInterceptor/MessageInterceptorTests.cs:
```diff
@@ -0,0 +1,79 @@
+
+    [Fact]
+    [OuterLoop]
+    [ActiveIssue(1170)]
```
It would be good to understand why this test fails before we disable it.   @khdang , can this test be changed to pass, or do we understand the failure enough to know it is a product issue?

roncain commented:
@dotnet-bot test outerloop Windows_NT
@dotnet-bot test outerloop selfhost Windows_NT

roncain commented:
@khdang -- would you please ping this PR when it is ready to review?  It still fails to build, which is why you get a 100% failure rate in OuterLoops.  You can see this by building SelfHostedWcfService.sln.  These same errors will occur when IIS attempts to build the contents of its virtual directory, meaning no WCF services are available.

```
Build FAILED.

"c:\OSS\wcf\src\System.Private.ServiceModel\tools\SelfHostedWcfService\SelfHostedWCFService.sln" (default target) (1) ->
"c:\OSS\wcf\src\System.Private.ServiceModel\tools\SelfHostedWcfService\SelfHostedWCFService.csproj" (default target) (2) ->
(CoreCompile target) -> 
  c:\OSS\wcf\src\System.Private.ServiceModel\tools\IISHostedWcfService\App_code\AsyncResult.cs(121,17): error CS0103: The name 'callback' does not exist in the current context [c:\OSS\wcf\src\System.Private.ServiceModel\tools\SelfHostedWcfService\SelfHostedWCFService.csproj]
  c:\OSS\wcf\src\System.Private.ServiceModel\tools\IISHostedWcfService\App_code\AsyncResult.cs(150,29): error CS1061: 'TAsyncResult' does not contain a definition for 'endCalled' and no extension method 'endCalled' accepting a first argument of type 'TAsyncResult' could be found (are you missing a using directive or an assembly reference?) [c:\OSS\wcf\src\System.Private.ServiceModel\tools\SelfHostedWcfService\SelfHostedWCFService.csproj]
  c:\OSS\wcf\src\System.Private.ServiceModel\tools\IISHostedWcfService\App_code\AsyncResult.cs(155,25): error CS1061: 'TAsyncResult' does not contain a definition for 'endCalled' and no extension method 'endCalled' accepting a first argument of type 'TAsyncResult' could be found (are you missing a using directive or an assembly reference?) [c:\OSS\wcf\src\System.Private.ServiceModel\tools\SelfHostedWcfService\SelfHostedWCFService.csproj]
  c:\OSS\wcf\src\System.Private.ServiceModel\tools\IISHostedWcfService\App_code\AsyncResult.cs(157,30): error CS1061: 'TAsyncResult' does not contain a definition for 'isCompleted' and no extension method 'isCompleted' accepting a first argument of type 'TAsyncResult' could be found (are you missing a using directive or an assembly reference?) [c:\OSS\wcf\src\System.Private.ServiceModel\tools\SelfHostedWcfService\SelfHostedWCFService.csproj]
  c:\OSS\wcf\src\System.Private.ServiceModel\tools\IISHostedWcfService\App_code\AsyncResult.cs(162,29): error CS1061: 'TAsyncResult' does not contain a definition for 'manualResetEvent' and no extension method 'manualResetEvent' accepting a first argument of type 'TAsyncResult' could be found (are you missing a using directive or an assembly reference?) [c:\OSS\wcf\src\System.Private.ServiceModel\tools\SelfHostedWcfService\SelfHostedWCFService.csproj]
  c:\OSS\wcf\src\System.Private.ServiceModel\tools\IISHostedWcfService\App_code\AsyncResult.cs(164,29): error CS1061: 'TAsyncResult' does not contain a definition for 'manualResetEvent' and no extension method 'manualResetEvent' accepting a first argument of type 'TAsyncResult' could be found (are you missing a using directive or an assembly reference?) [c:\OSS\wcf\src\System.Private.ServiceModel\tools\SelfHostedWcfService\SelfHostedWCFService.csproj]
  c:\OSS\wcf\src\System.Private.ServiceModel\tools\IISHostedWcfService\App_code\AsyncResult.cs(167,29): error CS1061: 'TAsyncResult' does not contain a definition for 'exception' and no extension method 'exception' accepting a first argument of type 'TAsyncResult' could be found (are you missing a using directive or an assembly reference?) [c:\OSS\wcf\src\System.Private.ServiceModel\tools\SelfHostedWcfService\SelfHostedWCFService.csproj]
  c:\OSS\wcf\src\System.Private.ServiceModel\tools\IISHostedWcfService\App_code\AsyncResult.cs(169,35): error CS1061: 'TAsyncResult' does not contain a definition for 'exception' and no extension method 'exception' accepting a first argument of type 'TAsyncResult' could be found (are you missing a using directive or an assembly reference?) [c:\OSS\wcf\src\System.Private.ServiceModel\tools\SelfHostedWcfService\SelfHostedWCFService.csproj]

    0 Warning(s)
    8 Error(s)
```

khdang commented:
Initially I run StartWCFSelfHostedService.cmd and it went pass the "Building WCF Self hosted sevice.." step without error so I thought there is no issues with the build. But I did notice the errors in your previous comments. Let me resolve that.

khdang commented:
@dotnet-bot test outerloop Windows_NT
@dotnet-bot test outerloop selfhost Windows_NT

khdang commented:
I've fixed the errors when building SelfHostedWCFService. It's a bug in CodeFormatter that it doesn't rename all the occurrences of variables.

I've fixed the tests issue as well to open the channels properly. I think this should verify that the channel extensibility model works. 

As for #623 2nd task, the LayeredChannel doesn't seem to be covered with this test so that need to be done separately.

khdang commented:
It looks like I need to update the test service for the new tests.

```
Starting:    Extensibility.MessageInterceptor.Tests
14:04:14      MessageInterceptorTests.CustomBinding_Message_Interceptor [FAIL]
14:04:14         System.ServiceModel.EndpointNotFoundException : There was no endpoint listening at http://wcfcoresrv2.cloudapp.net/WcfService15//ChannelExtensibility.svc//ChannelExtensibility that could accept the message. This is often caused by an incorrect address or SOAP action. See InnerException, if present, for more details.
14:04:14         Stack Trace:
```

roncain commented on src/System.Private.ServiceModel/tools/SelfHostedWcfService/SelfHostedWCFService.csproj:
```diff
@@ -39,6 +39,7 @@
       <Private>True</Private>
     </Reference>
     <Reference Include="System" />
+    <Reference Include="System.configuration" />
```
Nit: System.Configuration (upper-case 'C")

roncain commented on src/System.Private.ServiceModel/tools/SelfHostedWcfService/Program.cs:
```diff
@@ -286,6 +286,10 @@ private static void Main()
             webSocketHttpsDuplexTextBufferedTestServiceHostServiceHost.Open();
 
+            Uri[] channelExtensibilityTestServiceHostbaseAddress = new Uri[] { new Uri(string.Format("{0}/ChannelExtensibility.svc", httpBaseAddress)) };
+            ChannelExtensibilityServiceHost channelExtensiblityTestServiceHostServiceHost = new ChannelExtensibilityServiceHost(typeof(WcfChannelExtensiblityService), channelExtensibilityTestServiceHostbaseAddress);
```
I may have missed seeing it, but the web.config under tools\IISHostedWcfService needs to be updated to add an entry for the new servicehost.  I predict the Windows OuterLoops against IIS will fail without it.  Ignore this message if I missed seeing that file.

khdang replied:
I'll update the web.config to include the new servicehost.

roncain commented on src/System.Private.ServiceModel/tools/IISHostedWcfService/App_code/Interceptingelement.cs:
```diff
@@ -0,0 +1,38 @@
+    /// Configuration class for InterceptingBindingElement. To make your InterceptingBindingElement
+    /// configurable, derive from InterceptingElementInterceptingElement and override CreateMessageInterceptor()
+    /// </summary>
+    public abstract class InterceptingElement : BindingElementExtensionElement
```
Nit: the filename casing does not match the class name (Interceptingelement.cs).

roncain commented:
LGTM subject to a couple things:
 - Get CI to pass Windows OuterLoops by updating the web.config as mentioned in a comment
 - Once that passes, ask to run all outerloops (look at other PR's that contain "test all outerloop " -- I don't want to make trigger it by accident :grinning:)
 - Consider renaming the one file to match the casing of the class name

Once we get an CI OuterLoop pass for all OSes, I'm fine merging this.

khdang commented:
@dotnet-bot test outerloop Windows_NT
@dotnet-bot test outerloop selfhost Windows_NT

khdang commented:
@dotnet-bot test all outerloop please

khdang commented:
@dotnet-bot test OuterLoop Selfhost Windows_NT Release
@dotnet-bot test OuterLoop Ubuntu14.04 Debug
@dotnet-bot test OuterLoop Ubuntu14.04 Release

roncain commented:
@khdang -- I think the Ubuntu and OSX OuterLoop failures are unrelated to your PR, and you should go ahead and merge.
Looking at the failures, it looks like the PRService is encountering problems attempting to initialize the build tools.  If I sync my Ubuntu VM to your PR and run exactly the same build script that PRService attempted (found in the console log for the failure in the Details link), I actually see the new MessageInspector test run and pass.  I think we need to investigate the PR machine separately to discover why it failed to init the build tools -- but don't block this PR on that.

/cc: @iamjasonp

khdang commented:
Thanks @roncain! I will squash the commits and merge.

khdang commented:
@dotnet-bot test all outerloop please

iamjasonp commented:
ACK - issues look transient to me

roncain commented:
LGTM. I recommend checking in asap @khdang -- today is the last day before we branch.
The one failing test is not related to this PR.  I think it is a flakey test and I will fix that separately.
